taichi/ir/statements.h

```diff
@@ -311,9 +316,23 @@ class PtrOffsetStmt : public Stmt {
   Stmt *origin{nullptr};
   Stmt *offset{nullptr};
 
-  PtrOffsetStmt(Stmt *origin, Stmt *offset) : origin(origin), offset(offset) {
-    element_type() = origin->cast<GlobalPtrStmt>()->ret_type;
-    TI_STMT_REG_FIELDS;
+  PtrOffsetStmt(Stmt *, Stmt *);
+
+  bool is_local_ptr() const {
+    if (origin->is<AllocaStmt>() || origin->is<GlobalTemporaryStmt>()) {
+      TI_ASSERT_INFO(origin->ret_type->is<TensorType>(),
+                     "PtrOffsetStmt can only be used for Alloca (TensorType).")
+    }
+    return origin->is<AllocaStmt>() || origin->is<GlobalTemporaryStmt>();
+  }
+
+  bool is_unlowered_global_ptr() const {
+    return origin->is<GlobalPtrStmt>();
+  }
+
+  bool is_lowered_global_ptr() const {
+    // TODO: Check this is the only case.
+    return origin->is<GetChStmt>();
```

Comment:
Feels like a pretty restrictive assumption. Let's just use `!is_unlowered_global_ptr()` for now.

Reply:
Do you mean ```!is_local_ptr() && !is_unlowered_global_ptr()```?

Reply:
yeah, something like that

Reply:
https://github.com/taichi-dev/taichi/pull/2637/commits/aee940856346702e2dafd47ebb1636afbae29801